shared-bindings/bitmapfilter/__init__.c

```diff
@@ -0,0 +1,432 @@
+    if (args[ARG_mask].u_obj != mp_const_none) {
+        mp_arg_validate_type(args[ARG_mask].u_obj, &displayio_bitmap_type, MP_QSTR_mask);
+        mask = MP_OBJ_TO_PTR(args[ARG_mask].u_obj);
+    }
+
+    mp_float_t b = mp_obj_get_float(args[ARG_add].u_obj);
+
+    mp_obj_t weights = args[ARG_weights].u_obj;
+    mp_obj_t obj_len = mp_obj_len(weights);
+    if (obj_len == MP_OBJ_NULL || !mp_obj_is_small_int(obj_len)) {
+        mp_raise_ValueError_varg(MP_ERROR_TEXT("%q must be of type %q, not %q"), MP_QSTR_weights, MP_QSTR_Sequence, mp_obj_get_type(weights)->name);
+    }
+
+    size_t n_weights = MP_OBJ_SMALL_INT_VALUE(obj_len);
+
+    size_t sq_n_weights = (int)MICROPY_FLOAT_C_FUN(sqrt)(n_weights);
+    if (sq_n_weights % 2 == 0 || sq_n_weights * sq_n_weights != n_weights) {
+        mp_raise_ValueError(MP_ERROR_TEXT("weights must be a sequence with an odd square number of elements (usually 9 or 25)"));
+    }
+
+    int iweights[n_weights];
+    int weight_sum = 0;
+    for (size_t i = 0; i < n_weights; i++) {
+        mp_int_t j = mp_obj_get_int(mp_obj_subscr(weights, MP_OBJ_NEW_SMALL_INT(i), MP_OBJ_SENTINEL));
+        iweights[i] = j;
+        weight_sum += j;
+    }
+
+    mp_float_t m = get_m(args[ARG_mul].u_obj, weight_sum);
+
+    shared_module_bitmapfilter_morph(bitmap, mask, sq_n_weights / 2, iweights, m, b,
+        args[ARG_threshold].u_bool, args[ARG_offset].u_bool, args[ARG_invert].u_bool);
+    return args[ARG_bitmap].u_obj;
+}
+MP_DEFINE_CONST_FUN_OBJ_KW(bitmapfilter_morph_obj, 0, bitmapfilter_morph);
+static mp_obj_t subscr(mp_obj_t o, int i) {
+    return mp_obj_subscr(o, MP_OBJ_NEW_SMALL_INT(i), MP_OBJ_SENTINEL);
+}
+
+static mp_float_t float_subscr(mp_obj_t o, int i) {
+    return mp_obj_get_float(subscr(o, i));
+}
+
+//| def mix(
+//|     bitmap: displayio.Bitmap, weights: Sequence[int], mask: displayio.Bitmap | None = None
+//| ) -> displayio.Bitmap:
+//|     """Perform a channel mixing operation on the bitmap
+//|
+//|     The ``bitmap``, which must be in RGB565_SWAPPED format, is modified
+//|     according to the ``weights``.
+//|
+//|     If ``weights`` is a list of length 3, then each channel is scaled independently:
+//|     The numbers are the red, green, and blue channel scales.
+//|
+//|     If ``weights`` is a list of length 6, then each channel is scaled and
+//|     offset independently: The first two numbers are applied to the red channel:
+//|     scale and offset. The second two number are applied to the green
+//|     channel, and the last two numbers to the blue channel.
+//|
+//|     If ``weights`` is a list of length 9, then channels are mixed. The first three
+//|     numbers are the fraction of red, green and blue input channels mixed into the
+//|     red output channel. The next 3 numbers are for green, and the final 3 are for blue.
+//|
+//|     If ``weights`` is a list of length 12, then channels are mixed with an offset.
+//|     Every fourth value is the offset value.
```

Comment:
I wonder if this input would be clearer by using a named tuple for each color channel. That way you'd have a name for each number.

Reply:
I added some named tuple types for this purpose, and improved the documentation